docs(core-concepts): re-work architecture and navigation article (#1450)
NativeScript/docs: docs/core-concepts/application-architecture.md

---
title: Application Architecture
description: Learn the core concepts of building NativeScript apps
position: 30
slug: architecture
environment: nativescript
previous_url: /architecture
---

# Application Architecture

> In this article we are going to go through the core concepts of the core NativeScript framework.

Reviewer: This sounds too generic and claims to cover the whole "Core Concept" section. I would try to make it a bit more specific:

```suggestion
In this article we are going to review the architecture of a NativeScript application built with the Core Framework. It will cover the entry point of the app, the decomposition into modules, styling the application and data binding.
```

A root module can have only one component at the root of its content. You can put virtually any UI component as a root, but the most commonly used components are the one that can have children - the layouts, `TabView`, `SideDrawer` or `Frame`. The `Frame` component can't have children, but it can display and navigate between page modules.

Note that the root module will be loaded regardless of navigations until its UI container disappears. This basically means that the app root module will always be loaded. A modal view root module will be unloaded when the modal view is closed.

Here is an example of an app root module:

``` XML
<!-- app-root.xml -->
<Frame loaded="onFrameLoaded" defaultPage="home-page" />
```
``` JavaScript
// app-root.js
function onFrameLoaded(args) {
    console.log("Frame Loaded");
}
exports.onFrameLoaded = onFrameLoaded;
```
``` TypeScript
// app-root.ts
import { EventData } from "tns-core-modules/data/observable";

export function onFrameLoaded(args: EventData): void {
    console.log("Frame Loaded");
}
```

### Page Modules

> These modules represent pages and are used by the `Frame` component to implement forward na backward navigation. You can pass these modules to the `Frame` in one of two ways:

Reviewer 2: `forward na` typo

Reviewer: ```suggestion
These modules represent pages and are used by the `Frame` component to implement forward and backward navigation. You can pass these modules to the `Frame` in one of two ways:
```

* The `defaultPage` attribute - the page module set in this attribute will be initially shown by the `Frame`.
* The `navigate()` method - this method forces the `Frame` to hide the currently navigated page module and to show the page module passed as parameter.

Navigation is covered in detail in the [navigation article]({% slug navigation %}).

Page modules must always have the `Page` component at the root of their content. Below is a code sample of a page module. 

``` XML
<!-- home-page.xml-->
<Page class="page" loaded="onPageLoaded">
    <StackLayout>
        <Label text="Hooray! Home Page loaded!"/>
    </StackLayout>
</Page>
```
``` JavaScript
// home-page.js
function onPageLoaded(args) {
    console.log("Page Loaded");
}
exports.onPageLoaded = onPageLoaded;
```
``` TypeScript
// home-page.ts
import { EventData } from "tns-core-modules/data/observable";

export function onPageLoaded(args: EventData): void {
    console.log("Page Loaded");
}
```
``` CSS
/* home-page.css */
.page {
    background-color: teal;
}
```

## Global App Styling

> The core NativeScript framework also provides a way to set application wide styling. The default place to do that is in the `app.css` file in the app root folder. All css rules that are declared in this file will be applied to all application modules.

Reviewer: On the homepage the core framework is referred as "NativeScript Core". Should we stick to this naming? According to me, it looks more like a name of something.

Author: ```suggestion
The NativeScript Core framework also provides a way to set application wide styling. The default place to do that is in the `app.css` file in the app root folder. All css rules that are declared in this file will be applied to all application modules.
```

You can change the name of the file from which the application-wide CSS is loaded. You need to do the change before the `application.run()` method is called as shown below:

``` JavaScript
var application = require("tns-core-modules/application");
application.setCssFileName("style.css");

application.run({ moduleName: "main-page" });
```
``` TypeScript
import * as application from "tns-core-modules/application";
application.setCssFileName("style.css");

application.run({ moduleName: "main-page" });
```

Styling is covered in detail in the [styling article]({% slug styling %}).

## Supporting Multiple Screens

> Mobile applications are running on different devices with different screen sizes and form factors. NativeScript provides a way to define different files (.js, .css, .xml, etc.) to be loaded based on the screen's size, platform and orientation of the current device. The approach is somewhat similar to [multi screen support in Android](http://developer.android.com/guide/practices/screens_support.html). There is a set of *qualifiers* that can be added inside the file that will be respected when the file is loaded. Here is how the file should look:

Reviewer: Oxford comma

```suggestion
Mobile applications are running on different devices with different screen sizes and form factors. NativeScript provides a way to define different files (.js, .css, .xml, etc.) to be loaded based on the screen's size, platform, and orientation of the current device. The approach is somewhat similar to [multi screen support in Android](http://developer.android.com/guide/practices/screens_support.html). There is a set of *qualifiers* that can be added inside the file that will be respected when the file is loaded. Here is how the file should look:
```

`<file-name>[.<qualifier>]*.<extension>`

In the next section, we will go through the list of supported qualifiers.

### Screen Size Qualifiers

> All the values in screen size qualifiers are in density independent pixels(DP) &mdash; meaning it corresponds to the physical dimensions of the screen. The assumptions are that there are ~160 DP per inch. For example, according to Android guidelines, if the device's smaller dimension is more than 600 dp (~3.75 inches), it is probably a tablet.

Reviewer: Space before the opening bracket
```suggestion
All the values in screen size qualifiers are in density independent pixels (DP) &mdash; meaning it corresponds to the physical dimensions of the screen. The assumptions are that there are ~160 DP per inch. For example, according to Android guidelines, if the device's smaller dimension is more than 600 dp (~3.75 inches), it is probably a tablet.
```
